lib/Sema/CSSimplify.cpp

```diff
@@ -3449,24 +3448,95 @@ ConstraintSystem::simplifyMemberConstraint(ConstraintKind kind,
-    // a force unwrap fixit.
-    bool optionalChain;
-    if (!getContextualType())
-      optionalChain = !(Options & ConstraintSystemFlags::PreferForceUnwrapToOptional);
-    else
-      optionalChain = !getContextualType()->getOptionalObjectType().isNull();
-    auto fixKind = optionalChain ? FixKind::OptionalChaining : FixKind::ForceOptional;
-
-    // Note the fix.
-    if (recordFix(fixKind, locator))
-      return SolutionKind::Error;
-    
+
+    // Figure out if we have an existing optional chaining fix done earlier in
+    // this dotted expression chain.
+    bool continuingChain = false;
+    if (auto dotExpr = dyn_cast<UnresolvedDotExpr>(locator->getAnchor())) {
+      for (auto pair : Fixes) {
+        if (pair.first.getKind() == FixKind::OptionalChaining) {
+          auto innerExpr = dotExpr->getBase();
+          while (innerExpr) {
+            if (pair.second->getAnchor() == innerExpr) {
+              continuingChain = true;
+              break;
+            }
+            if (auto innerDot = dyn_cast<UnresolvedDotExpr>(innerExpr))
+              innerExpr = innerDot->getBase();
+          }
+        }
+      }
+    }
+
+    // If we aren't continuing an existing chain, do we want to add a new fixit?
+    if (!continuingChain) {
+      if (!shouldAttemptFixes())
+        return SolutionKind::Error;
+
+      auto contextTy = getContextualType();
+      bool forceUnwrap;
+      if (contextTy)
+        forceUnwrap = contextTy->getOptionalObjectType().isNull();
+      else
+        forceUnwrap = Options.contains(ConstraintSystemFlags::PreferForceUnwrapToOptional);
+
+      // Note the fix.
+      auto fixKind = forceUnwrap ? FixKind::ForceOptional : FixKind::OptionalChaining;
+      if (recordFix(fixKind, locator))
+        return SolutionKind::Error;
+
+      if (forceUnwrap) {
+        // The simple case - just look through one level of optional.
+        addValueMemberConstraint(unwrappedBaseObjTy,
+                                 member, memberTy, useDC, functionRefKind, locator);
+        return SolutionKind::Solved;
+      }
+    }
+
+    // See if there is a matching member with an optional type.
+    // If so, then inserting an optional chaining fix won't produce a MORE optional result,
+    // and it is safe to just fix with chaining.
+    MemberLookupResult result =
+      performMemberLookup(kind, member, unwrappedBaseObjTy, functionRefKind, locator,
+                          /*includeInaccessibleMembers*/false);
+    bool memberIsOptional = false;
+    for (auto candidate : result.ViableCandidates) {
+      if (candidate.isDecl())
+        memberIsOptional |= !candidate.getDecl()->getInterfaceType()->
+                              getOptionalObjectType().isNull();
+    }
+
+    // If the member isn't already optional, we need to make the result optional.
+    if (!memberIsOptional) {
+      auto innerTV = createTypeVariable(locator, TVO_CanBindToLValue | TVO_CanBindToInOut);
```

Comment:
Can this really bind to `inout`?

Reply:
This is cribbed from how the type variables for non-implicit optional chaining are created in CSGen. I can look at whether it's necessary or not.

Reply:
“Can bind to inout” used to be the default. When I made an explicit flag for it, I went and added it in every place we were creating a type variable. Most of them can lose the flag - that was the original motivation for adding it - but nobody got around to cleaning that up yet. Maybe @CodaFi can do it :)